### ISSUE TITLE
Kubernetes kata-fc & kata-qemu setup with CRI-O

### ISSUE THREAD

**Issue by joost14** [kata-containers/documentation#389]
Hello, 
Im a bit new when it comes to Kubernetes and container runtimes. Im currently trying to setup a Kubernets cluster using kubeadm with crio v1.13. I'm trying to select kata-qemu or kata-fc as the runtime for one of my pods. As selecting untrusted workloads in crio.conf will be removed in v1.14 it cannot be using untrusted workload as showed within the following guide: 
https://github.com/kata-containers/documentation/blob/master/Developer-Guide.md#run-kata-containers-with-kubernetes.  

I've installed kata for ubuntu using the follwoing guide: 
https://github.com/kata-containers/documentation/blob/master/install/ubuntu-installation-guide.md 

I tried using RuntimeClass using this guide: 
https://katacontainers.io/posts/kata-deploy-install-kata-containers-kubernetes/ 
When applying the following patch to update the runtime spec of my busybox deployment nothing changes : 
   
spec:
      template:
      spec:
          runtimeClassName: kata-qemu

Used command to patch: kubectl patch deploy busybox --patch "$(cat kata-qemu-patch.yaml)"
output: deployment.extensions/busybox patched (no changes). 
This is the same when using kata-fc
My crio.conf ( I believe I'm missing something here ): 
  [crio.runtime.runtimes.runc]
  runtime_path = "/usr/lib/cri-o-runc/sbin/runc"

  [crio.runtime.runtimes.kata]
  runtime_path = "/usr/bin/kata-runtime"

And I've tried kata-deploy (not using he kata install guide for ubuntu) :
https://github.com/clearlinux/cloud-native-setup/tree/master/clr-k8s-examples/8-kata 
When using kata-deploy it says it installs kata and configures it automatically, but when running the command kata-runtime version it says command not found. 

Crio version: 
1.13.1-dev

Kata version:
v1.6.0-rc1

In short, when using kata deploy it seems it will not install kata-runtime. And when using the ubuntu installation guide it seems im not specifying kata-qemu and kata-fc in the crio.conf correctly as I have no idea how. 
Which leaves me with the follwoing questions: 
- How do you specify kata-qemu and kata-fc in the crio.conf?
- Is firecracker installed when using the Ubuntu installation guide?
- How do you configure RuntimeClass correctly to switch container runtimes within pods?

**Comment by chavafg** [comment]
Ho @joost14. 

1. If you want to use `RuntimeClass`, you should add `kata-runtime` to the runtimes map in the crio.conf file. Here is what we use for our CI: https://github.com/kata-containers/tests/blob/master/.ci/configure_crio_for_kata.sh#L19-L25

2. You should also enable the runtimeclass feture on kubelet and apiserver (maybe you already did this) as mentioned on https://kubernetes.io/docs/concepts/containers/runtime-class/ 

3. Here is one example of the an nginx deployment using runtimeclass with kata. 

Hope that this helps you.  If you have any further question, please let us know. 
And yes, we need to update our documentation for using `RuntimeClass`

**Comment by joost14** [comment]
Hi @chavafg 

Thank you for your reply. 

- I've added the following to the crio.conf 

  [crio.runtime.runtimes.kata]
  runtime_path = "/usr/bin/kata-runtime"

- I've already enabled the runtimeclass future as mentioned. Do you have an example how to specify the runtime handler? Right now i am using the following: 

kind: RuntimeClass
apiVersion: node.k8s.io/v1alpha1
metadata:
    name: kata
spec:
    runtimeHandler: kata

- What example do you mean? I see no link...

**Comment by chavafg** [comment]
Sorry, forgot to add the link:
This is our handler:
https://github.com/kata-containers/tests/blob/master/integration/kubernetes/runtimeclass_workloads/kata-runtimeclass.yaml

And examples of the workloads using `runtimeclass` can be found here:
https://github.com/kata-containers/tests/tree/master/integration/kubernetes/runtimeclass_workloads

**Comment by joost14** [comment]
Hey @chavafg, 

I've tried your way of specifying the handler in crio.conf, but still no luck. I see you have the runtime installed in the local map. Mine is in /usr/bin/kata-runtime. How did you install the runtime and the hypervisor (qemu & firecracker)? 

I followed the ubuntu guide in the kata-runtime docs: 
https://github.com/kata-containers/documentation/blob/master/install/ubuntu-installation-guide.md

**Comment by chavafg** [comment]
Hi @joost14, 

The scripts above are used in the CI and build each component from sources, that is why the binaries are under `/usr/local`. If you followed the ubuntu guide, it should be fine.

Can you paste here your whole `crio.conf` file and the `kubelet` systemd file (under `/etc/systemd/system/kubelet.service.d/`) here? 
Also, can you check the kubelet log (`sudo journalctl -u kubelet`) and search for errors there? Please if you find errors there, paste them here.

**Comment by egernst** [comment]
Kata deploy installs the binaries at /opt/kata/bin and points the crio configuration to this path. I’m guessing your crio conf might have issues at this point. 

Can you share it?

**Comment by joost14** [comment]
```
[crio]
file_locking = false
file_locking_path = "/run/crio.lock"

[crio.api]
listen = "/var/run/crio/crio.sock"
stream_address = "127.0.0.1"
stream_port = "0"
stream_enable_tls = false
stream_tls_cert = ""
stream_tls_key = ""
stream_tls_ca = ""
grpc_max_send_msg_size = 16777216
grpc_max_recv_msg_size = 16777216

[crio.runtime]
default_runtime = "runc"
no_pivot = false
conmon = "/usr/lib/crio/bin/conmon"
conmon_env = [
	"PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin",
]
selinux = false
seccomp_profile = "/etc/crio/seccomp.json"
apparmor_profile = "crio-default"
cgroup_manager = "systemd"
default_capabilities = [
	"CHOWN", 
	"DAC_OVERRIDE", 
	"FSETID", 
	"FOWNER", 
	"NET_RAW", 
	"SETGID", 
	"SETUID", 
	"SETPCAP", 
	"NET_BIND_SERVICE", 
	"SYS_CHROOT", 
	"KILL", 
]
default_sysctls = [
]
additional_devices = [
]
hooks_dir = [
]
default_mounts = [
]

pids_limit = 1024
log_size_max = -1
container_exits_dir = "/var/run/crio/exits"
container_attach_socket_dir = "/var/run/crio"
read_only = false
log_level = "error"
uid_mappings = ""
gid_mappings = ""
ctr_stop_timeout = 0
  
  [crio.runtime.runtimes.runc]
  runtime_path = "/usr/lib/cri-o-runc/sbin/runc"
  
  [crio.runtime.runtimes.kata]
  runtime_path = "/usr/bin/kata-runtime"

[crio.image]
default_transport = "docker://"
pause_image = "k8s.gcr.io/pause:3.1"
pause_command = "/pause"
signature_policy = ""
image_volumes = "mkdir"
registries = [
 	"quay.io",
	"docker.io",
]


[crio.network]
network_dir = "/etc/cni/net.d/"
plugin_dir = "/opt/cni/bin"
```

I'm not using kata-deploy at the moment. I've only installed kata using the Ubuntu installation guide. 
I'm using kubeadm when deploying my cluster. So maybe it has to do with the way i add the feature-flag to the deployement. I'm using `kubeadm init  --cri-socket=”var/run/crio/crio.sock --config config.yaml `

config.yaml:
```
apiVersion: kubeadm.k8s.io/v1alpha3
kind: InitConfiguration
nodeRegistration:
  kubeletExtraArgs:
    "v": "2"
    "feature-gates": "RuntimeClass=true"
```

and the 10-kubeadm-conf:
```
# Note: This dropin only works with kubeadm and kubelet v1.11+
[Service]
Environment="KUBELET_KUBECONFIG_ARGS=--bootstrap-kubeconfig=/etc/kubernetes/bo$
Environment="KUBELET_CONFIG_ARGS=--config=/var/lib/kubelet/config.yaml"
# This is a file that "kubeadm init" and "kubeadm join" generates at runtime, $
EnvironmentFile=-/var/lib/kubelet/kubeadm-flags.env
# This is a file that the user can use for overrides of the kubelet args as a $
# the .NodeRegistration.KubeletExtraArgs object in the configuration files ins$
EnvironmentFile=-/etc/default/kubelet
ExecStart=
ExecStart=/usr/bin/kubelet $KUBELET_KUBECONFIG_ARGS $KUBELET_CONFIG_ARGS $KUBE$
```
and the kubeadm-flags.env
`KUBELET_KUBEADM_ARGS=--feature-gates=RuntimeClass=true --container-runtime=remote --container-runtime-endpoint=unix:///var/run/crio/crio.sock --resolv-conf=/run/systemd/resolve/resolv.conf`

**Comment by chavafg** [comment]
Above, I don't see if you are enabling the `runtimeclass` to the apiserver? 
I do this on the same config.yaml that is passed to the `kubeadm init`, this is what we use : 
```
apiVersion: kubeadm.k8s.io/v1alpha3
kind: InitConfiguration
nodeRegistration:
  criSocket: /var/run/crio/crio.sock
---
apiVersion: kubeadm.k8s.io/v1alpha3
kind: ClusterConfiguration
kubernetesVersion: v1.13.3
networking:
  dnsDomain: cluster.local
  podSubnet: 10.244.0.0/16
  serviceSubnet: 10.96.0.0/12
apiServerExtraArgs:
  feature-gates: RuntimeClass=true

```

**Comment by joost14** [comment]
Hi @chavafg 

The RuntimeClass feature-gate seems to be working now, but the pod still does not switch the runtime. I tested this with busybox. When deploying with and without the runtimeClassName i get the same kernel as the system its running on. Is this normal or should I use an other install than the Ubuntu guide?: https://github.com/kata-containers/documentation/blob/master/install/ubuntu-installation-guide.md 
I believe the Ubuntu guide does not install firecracker, is this true?  

I'm deploying busybox using the following template: 
```
apiVersion: v1
kind: Pod
metadata:
  name: busybox
spec:
  runtimeClassName: kata
  shareProcessNamespace: true
  containers:
  - name: busybox
    image: busybox
    command:
      - sleep
- "120"
```

**Comment by chavafg** [comment]
Hi @joost14 
> When deploying with and without the runtimeClassName i get the same kernel as the system its running on. Is this normal or should I use an other install than the Ubuntu guide?

Not normal, seems like still using runc. When running kata successfully, you should see another kernel version. 

I think this is still a configuration issue in k8s.

Can you verify if your the name of the runtime specified on crio.conf matches the one specified in the k8s runtime handler?
E.g: if 
```
[crio.runtime.runtimes.kata-qemu]
  runtime_path = "/opt/kata/bin/kata-qemu"
```
then `kata-qemu` should be specified in the runtimeclass runtimehandler:
```

kind: RuntimeClass
apiVersion: node.k8s.io/v1alpha1
metadata:
  name: kata
spec:
  runtimeHandler: kata-qemu
```

Also, I am not sure if firecracker works with cri-o and k8s, @egernst do you know if this is already working? 


Last thing you may want to try is using our CI k8s scripts and see if you can make it work... 
If you already have the cri-o and kata setup, you can:

```
# clone our tests repository: 
go get github.com/kata-containers/tests
cd $GOPATH/src/github.com/kata-containers/tests
export use_runtime_class=true
# run k8s install: 
.ci/install_kubernetes.sh

cd $GOPATH/src/github.com/kata-containers/tests/integration/kubernetes
# run init script 
./init.sh 
```
Our scripts assume that kata is configured in crio.conf as follows:
```
[crio.runtime.runtimes.kata]
  runtime_path = "/PATH/TO/YOUR/KATA_BINARY"
```

**Comment by joost14** [comment]
Hi @chavafg 

Yes I've specified ` kata` as handler and the crio.conf looks like this (the crio.conf is different now as in the previous post I used kata-deploy not the ubuntu install of kata): 
```
[crio.runtime.runtimes.kata]
runtime_path = "/usr/bin/kata-runtime"
```

I've tried your CI k8s scripts, but when running the k8s install `.ci/install_kubernetes.sh` i get the following line: 
`.ci/install_kubernetes.sh: line 35: chronic: command not found`

And when runing the init script `./init.sh`:
`/home/test1/go/src/github.com/kata-containers/tests/integration/kubernetes/../../lib/common.bash: line 57: RUNTIME: unbound variable
INFO: Runtime environment not found - setting defaults
/home/test1/go/src/github.com/kata-containers/tests/integration/kubernetes/../../lib/common.bash: line 104: RUNTIME_PATH: unbound variable`

I've used a ubuntu 18.04.2 machine with crio and kata installed when running the scripts.

**Comment by jodh-intel** [comment]
@joost14 - Hi, To install `chronic` on Ubuntu:

```sh
$ sudo apt -y install moreutils
```

(Our CI does that here: https://github.com/kata-containers/tests/blob/master/.ci/setup_env_ubuntu.sh#L17..L18)

Then, you'll need to set the two variables mentioned:

```sh
export RUNTIME=kata-runtime
export RUNTIME_PATH=/usr/bin/kata-runtime
```

(Our CI does that here: https://github.com/kata-containers/ci/blob/master/jenkins/jobs/kata-containers-containerd-k8s-conformance-e2e/config.xml#L45..L46).

**Comment by joost14** [comment]
@chavafg @jodh-intel @egernst  Hi, thanks for the help. It seems to be working now. 

After installing the script and deploying and patching a container on a single node the kernel changed. Which indicates kata is being used.

I noticed when I joined a worker node the kernel did not change. The problem was when joining the worker node the feature gate flag was not set correctly. After adding the the feature-gate in the kubeadm.conf it worked.

I added the following in `/etc/systemd/system/kubelet.service.d/10-kubeadm.conf`:
`ExexStart= --feature-gates RuntimeClass=true`
And joined the node. 

I rebuild my master and worker (without the script) and it seems to be working fine now. 

Thank you for all the help :+1:

**Comment by chavafg** [comment]
Hi @joost14, good to know that it is now working. 👍

**Comment by jodh-intel** [comment]
Hi @joost14 - are you happy to close this now the problem has been resolved?

**Comment by joost14** [comment]
Yes
